### ISSUE TITLE
spring-cloud-gcp-starter-pubsub doesn't work with spring-cloud-gcp-starter-trace version higher than 1.0.0.RC

### ISSUE THREAD

**Issue by shiraji** [spring-cloud/spring-cloud-gcp#1373]
Hi, there

We have been struggling with spring-cloud-gcp-starter-trace.

If we include the dependency of spring-cloud-gcp-starter-trace version higher than 1.0.0.RC, the app does not receive any Cloud Pub/Sub messages.

Here is the minimum sample project that reproduce this problem. https://github.com/shiraji/demo-pubsub

This project does not work with Cloud Pub/Sub in version 1.1.0.RC2. However, once we remove the dependency of spring-cloud-gcp-starter-trace, this project works perfectly.
Also, if we use 1.0.0.RC for all spring-cloud-gcp-starter's, the project works nicely. 

It may be related to the problem that spring-cloud-gcp-starter-trace version 1.0.0.RELEASE does not report to StackDriver Trace. (1.0.0.RC1 works fine)

Thanks in advance.

**Comment by meltsufin** [comment]
Hi @shiraji,
Thanks for reporting the issue and providing the repro project. We will take a look.
In the meantime, we've recently updated to the latest versions of the Google Cloud Java libraries. Have you tried the same with the latest `1.1.0.BUILD-SNAPSHOT`? I wonder if it might contain a fix.

**Comment by shiraji** [comment]
I thought 1.1.0.RC2 is highest. Let me try.

**Comment by meltsufin** [comment]
It's in snapshot repo. https://repo.spring.io/snapshot/org/springframework/cloud/spring-cloud-gcp/

**Comment by dzou** [comment]
Hi shiraji,

The problem appears to be an issue with our dependency on grpc; the trace starter pulls in a different version which appears to cause the issue.

As a temporary workaround, you can modify your gradle file by adding an `exclude group` block to the trace starter dependency in your gradle file. With this modification, I was able to successfully send/receive messages through the sample application:

```
compile("org.springframework.cloud:spring-cloud-gcp-starter-trace:$springCloudGcpVersion") {
  exclude group: 'io.grpc'
}
```

We will investigate a permanent solution for our new release - it is possible that that the `1.1.0.BUILD-SNAPSHOT` does not have this problem to begin with. If you tried it out with the latest version, let us know if you ran into any problems.

**Comment by meltsufin** [comment]
It looks like the grpc `1.17.0` version is coming from `io.zipkin.gcp:zipkin-sender-stackdriver:0.9.0`.
I doesn't take precedence with Maven, but does with Gradle for some reason.

**Comment by suztomo** [comment]
@shiraji I ran your demo-pubsub project and got following error. Would you confirm this is the problem you observed?

```
demo-pubsub $ ./gradlew build

> Task :test

com.github.shiraji.demopubsub.DemoPubsubApplicationTests > contextLoads FAILED
    java.lang.IllegalStateException
        Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException
            Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException
                Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException
                    Caused by: org.springframework.beans.factory.BeanCreationException
                        Caused by: org.springframework.beans.BeanInstantiationException
                            Caused by: java.lang.IllegalArgumentException

1 test completed, 1 failed

> Task :test FAILED

FAILURE: Build failed with an exception.
```

**Comment by shiraji** [comment]
Hi, guys

Unfortunately, I observed the same behavior with 1.1.0.BUILD-SNAPSHOT.

@dzou 's idea which adding `exclude group: 'io.grpc'` is the only way to fix this problem at this moment.

@suztomo I'm sorry but I didn't touch test cases. This test came from Spring Initializr.

**Comment by meltsufin** [comment]
I tried removing Spring Cloud GCP from the picture and just testing using Google Cloud Java Pub/Sub library directly. I'm able to reproduce the issue, when `grpc-netty-shaded` is `1.16.1` and `grpc-protobuf` and `grpc-stub` are `1.17.1`.
```
dependencies {
    compile group: 'com.google.cloud', name: 'google-cloud-pubsub', version: '1.59.0'
    compile group: 'io.grpc', name: 'grpc-netty-shaded', version: '1.16.1'
    compile group: 'io.grpc', name: 'grpc-protobuf', version: '1.17.1'
    compile group: 'io.grpc', name: 'grpc-stub', version: '1.17.1'
}
```

When I use the streaming pull, there is no error message anywhere, but when I use the synchronous pull, I get this:

```
PubsubGrcpTest > testPull FAILED
    com.google.api.gax.rpc.UnauthenticatedException: io.grpc.StatusRuntimeException: UNAUTHENTICATED: Credentials should use fail() instead of throwing exceptions

        Caused by:
        io.grpc.StatusRuntimeException: UNAUTHENTICATED: Credentials should use fail() instead of throwing exceptions

            Caused by:
            java.lang.NoSuchMethodError: io.grpc.internal.AbstractClientStream.<init>(Lio/grpc/internal/WritableBufferAllocator;Lio/grpc/internal/StatsTraceContext;Lio/grpc/internal/TransportTracer;Lio/grpc/Metadata;Z)V
```

**Comment by meltsufin** [comment]
Here's the code I used: https://github.com/meltsufin/pubsubgrpctest/tree/master

**Comment by meltsufin** [comment]
We figured out that the reason we have two different versions of grpc dependencies in the project is because Google Cloud Java is on GRPC 1.16 and [zipkin-gcp](https://github.com/openzipkin/zipkin-gcp) is on 1.17. So, we're trying to address this problem by excluded grpc deps from zipkin-gcp. See #1379.

**Comment by meltsufin** [comment]
Is the issue resolved? @dzou Does the OP's repro case pass now?

**Comment by dzou** [comment]
Yes, the repro case now works after the submitted changes.

@meltsufin - what's the recommend course of action for @shiraji now? Should he wait for the new release 1.1.0 to remove the grpc exclusion we suggested?

**Comment by meltsufin** [comment]
1.1.0.RELEASE is coming in a few days. I think we can close this issue.

**Comment by dzou** [comment]
Ok, marking the issue as resolved now then.

shiraji - Thanks again for reporting the issue. The fix will be in 1.1.0.RELEASE which is coming in a few days; once released, you can use the latest version and can remove that dep exclusion we suggested earlier.

**Comment by shiraji** [comment]
Thanks guys!!!
